### DISCUSSION TITLE
FreeBSD 13.0-RELEASE based experimental 0.7.0 build

### DISCUSSION THREAD

**Discussion by probonopd** [helloSystem/hello#229] [General]
![image](https://user-images.githubusercontent.com/2480569/138579870-6ca16337-0f6d-47b7-86a9-fdf6e3405df7.png)

https://github.com/helloSystem/ISO/releases/tag/experimental-13.0

Showstoppers:

https://github.com/helloSystem/ISO/issues/198

If we cannot solve these in time, 0.7.0 will continue to be 12.2 based.

**Comment by grahamperrin** [comment]
Spun off from https://github.com/helloSystem/ISO/issues/198#issuecomment-950254675

> 0G35 cannot boot on non-EFI systems: …

`hello-0.7.0_0G35-FreeBSD-13.0-amd64.iso` booting without EFI: 

![image](https://user-images.githubusercontent.com/192271/138610476-8bf8beb3-2cb6-42b0-8fd2-01de7ea050f2.png)

Desktop environment without EFI: 

![image](https://user-images.githubusercontent.com/192271/138610533-51d3671c-6d41-46f3-b0de-8ab38c574f2a.png)

**Comment by grahamperrin** [comment]
Spun off from https://github.com/helloSystem/ISO/issues/198#issuecomment-950253162

> Why is the 0.7.0 build ~500 MB larger with FreeBSD 13? …

[Accessing .uzip file content : freebsd](https://old.reddit.com/r/freebsd/comments/qf4jtl/accessing_uzip_file_content/) – I'm confused by uzip

**Reply by probonopd** [reply]
Me too. It is too hard to mount currently. We should make it double-clickable just like other archives. Someone "just" needs to script it, I guess.

**Reply by grahamperrin** [reply]
> too hard to mount

How _do_ you mount?

**Reply by probonopd** [reply]
https://github.com/helloSystem/ISO/blob/3a612449a57f468ca29eff25a318d3373ed88725/overlays/ramdisk/init.sh#L39-L42

**Reply by grahamperrin** [reply]
I mean, with reference to <https://old.reddit.com/r/freebsd/comments/qf4jtl/accessing_uzip_file_content/>, how can I mount i.e. work with `md0` and `md1`?

**Reply by probonopd** [reply]
I thought that it should work like this:

```
mdconfig -o readonly -f system.uzip -u 1
zpool import furybsd -o readonly=on
zpool list # furybsd
```

**Reply by grahamperrin** [reply]
```text
root@mowa219-gjp4-8570p-freebsd:~ # cd /Volumes/t500/VirtualBox/BSD/helloSystem/helloSystem\ 0.6.0/
root@mowa219-gjp4-8570p-freebsd:/Volumes/t500/VirtualBox/BSD/helloSystem/helloSystem 0.6.0 # mdconfig -o readonly -f system.uzip -u 1
root@mowa219-gjp4-8570p-freebsd:/Volumes/t500/VirtualBox/BSD/helloSystem/helloSystem 0.6.0 # zpool import -N furybsd
cannot import 'furybsd': no such pool available
root@mowa219-gjp4-8570p-freebsd:/Volumes/t500/VirtualBox/BSD/helloSystem/helloSystem 0.6.0 # 
```

**Reply by probonopd** [reply]
Let's see.

```
FreeBSD% file '/media/LIVE/data/system.uzip'      
/media/LIVE/data/system.uzip: POSIX shell script executable (binary data)
```

Interesting. There seems to be an embedded shell script. Let's look at it:

```
FreeBSD% head -n 4 '/media/LIVE/data/system.uzip' 
#!/bin/sh
#v3.0 Format
(kldstat -qm g_uzip||kldload geom_uzip)>&-&&mount_cd9660 /dev/`mdconfig -af $0`.uzip $1
exit $?
```

Can we run it?

```
FreeBSD% sudo sh -x '/media/LIVE/data/system.uzip' 
+ kldstat -qm g_uzip
+ kldload geom_uzip
+ mdconfig -af /media/LIVE/data/system.uzip
mdconfig: WARNING: opening backing store: /media/LIVE/data/system.uzip readonly
+ mount_cd9660 /dev/md0.uzip
usage: mount_cd9660 [-begjrv] [-C charset] [-o options] [-s startsector]
                    special node
+ exit 64
```

No. Why?

Likely because we don't really have an ISO9660 filesystem on that uzip, but zfs, and the embedded script can't handle that.

But `zpool list` now shows something:

```
FreeBSD% zpool list
NAME      SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
furybsd      -      -      -        -         -      -      -      -  UNAVAIL  -
livecd       -      -      -        -         -      -      -      -  UNAVAIL  -
(...)
```

helloSystem comes with a script that mounts all available zfs filesytems: `zfs-import-altroot.`

Let's run it:

```
FreeBSD% sudo zfs-import-altroot

FreeBSD% mount
(...)
zfs-altroot-14188878336022036172 on /media/zfs-altroot-14188878336022036172/usr/local/furybsd/uzip (zfs, local, read-only, nfsv4acls)

FreeBSD% ls /media/zfs-altroot-14188878336022036172/usr/local/furybsd/uzip
Applications    dev             libexec         proc            sys             usr
bin             etc             media           rescue          System          var
boot            home            mnt             root            tmp
COPYRIGHT       lib             net             sbin            Users

FreeBSD% cat /media/zfs-altroot-14188878336022036172/usr/local/furybsd/uzip/.url 
https://github.com/helloSystem/ISO/commit/0daeeb2
```

So yes, we are accessing the contents of the correct uzip here.

To "unmount":

```
FreeBSD% sudo zfs-export-altroot
```

Now the zfs is no longer mounted.

How does one properly "unattach" the uzip from `/dev/mdX`?

**Reply by grahamperrin** [reply]
My previous attempt omitted `geom_uzip`. 

```text
# kldload geom_uzip
# cd /Volumes/t500/VirtualBox/BSD/helloSystem/helloSystem\ 0.6.0/
# mdconfig -o readonly -f system.uzip -u 6
# cd ../helloSystem\ 0.7.0/
# mdconfig -o readonly -f system.uzip -u 7
# mdconfig -l
md6 md7 
# zpool import 
   pool: furybsd
     id: 11467834712836977993
  state: ONLINE
status: Some supported features are not enabled on the pool.
        (Note that they may be intentionally disabled if the
        'compatibility' property is set.)
 action: The pool can be imported using its name or numeric identifier, though
        some features will not be available without an explicit 'zpool upgrade'.
 config:

        furybsd       ONLINE
          md6.uzipp1  ONLINE

   pool: furybsd
     id: 5751110427541885720
  state: ONLINE
 action: The pool can be imported using its name or numeric identifier.
 config:

        furybsd       ONLINE
          md7.uzipp1  ONLINE
# mkdir -p /tmp/0.6.0
# mkdir -p /tmp/0.7.0
# zpool import -R /tmp/0.6.0 -o readonly=on 11467834712836977993 h060
# zpool import -R /tmp/0.7.0 -o readonly=on 5751110427541885720 h070
# zpool status h060
  pool: h060
 state: ONLINE
status: Some supported and requested features are not enabled on the pool.
        The pool can still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(7) for details.
config:

        NAME          STATE     READ WRITE CKSUM
        h060          ONLINE       0     0     0
          md6.uzipp1  ONLINE       0     0     0

errors: No known data errors
# zpool status h070
  pool: h070
 state: ONLINE
config:

        NAME          STATE     READ WRITE CKSUM
        h070          ONLINE       0     0     0
          md7.uzipp1  ONLINE       0     0     0

errors: No known data errors
# time du -hs /tmp/0.6.0
1.5G    /tmp/0.6.0
       17.12 real         0.12 user         1.30 sys
# time du -hs /tmp/0.7.0
1.4G    /tmp/0.7.0
       25.08 real         0.07 user         1.42 sys
# ls -hl ../helloSystem\ 0.6.0/system.uzip 
-rwxr-xr-x  1 grahamperrin  grahamperrin   1.4G Sep 30 09:54 ../helloSystem 0.6.0/system.uzip
# du -h ../helloSystem\ 0.6.0/system.uzip
1.4G    ../helloSystem 0.6.0/system.uzip
# ls -hl system.uzip
-rwxr-xr-x  1 grahamperrin  grahamperrin   1.9G Oct 24 04:18 system.uzip
# du -h system.uzip
1.9G    system.uzip
# 
```

Whilst one of the two `system.uzip` files is considerably larger, for the associated file systems the sums of sizes (measured by Filelight) barely differ. 2.7 GiB in both cases: 

![image](https://user-images.githubusercontent.com/192271/139587211-b95154c3-12a0-4e40-bcd1-bb6d11e14c0c.png)

![image](https://user-images.githubusercontent.com/192271/139587269-71af8d40-ef1b-4ea5-a93d-cf7860f49590.png)

Housekeeping, after browsing the contents of the pools: 

```text
# zpool export h060
# zpool export h070
# mdconfig -d -u md6
# mdconfig -d -u md7
# 
```

**Reply by probonopd** [reply]
Very interesting investigation! What could be causing this difference?

**Reply by grahamperrin** [reply]
An uneducated guess: with FreeBSD 13.0⋯, maybe change your use of [mkuzip(8)](https://www.freebsd.org/cgi/man.cgi?query=mkuzip&sektion=8&manpath=FreeBSD) to `zstd-19`. 

NB: 

* <https://www.freebsd.org/cgi/man.cgi?query=mkuzip&sektion=8&manpath=FreeBSD#IMPLEMENTATION_NOTES>
* <https://www.freebsd.org/cgi/man.cgi?query=geom_uzip&sektion=4&manpath=FreeBSD#HISTORY>

[geom_uzip(4)](https://www.freebsd.org/cgi/man.cgi?query=geom_uzip&sektion=4&manpath=FreeBSD) 

----

```text
root@mowa219-gjp4-8570p-freebsd:~ # zfs get compression h060 h070
NAME  PROPERTY     VALUE           SOURCE
h060  compression  gzip            local
h070  compression  zstd-9          local
root@mowa219-gjp4-8570p-freebsd:~ # zfs get logicalused,logicalreferenced,refcompressratio,written,used,referenced,compressratio h060
NAME  PROPERTY           VALUE     SOURCE
h060  logicalused        4.46G     -
h060  logicalreferenced  4.46G     -
h060  refcompressratio   3.17x     -
h060  written            1.43G     -
h060  used               1.43G     -
h060  referenced         1.43G     -
h060  compressratio      3.17x     -
root@mowa219-gjp4-8570p-freebsd:~ # zfs get logicalused,logicalreferenced,refcompressratio,written,used,referenced,compressratio h070
NAME  PROPERTY           VALUE     SOURCE
h070  logicalused        4.57G     -
h070  logicalreferenced  4.57G     -
h070  refcompressratio   3.33x     -
h070  written            1.39G     -
h070  used               1.39G     -
h070  referenced         1.39G     -
h070  compressratio      3.33x     -
root@mowa219-gjp4-8570p-freebsd:~ # zpool export h060 h070
root@mowa219-gjp4-8570p-freebsd:~ # mdconfig -d -u md6
root@mowa219-gjp4-8570p-freebsd:~ # mdconfig -d -u md7
root@mowa219-gjp4-8570p-freebsd:~ # mdconfig -l
root@mowa219-gjp4-8570p-freebsd:~ # exit
logout
% 
```

**Reply by probonopd** [reply]
Am I reading it wrong? h070 has a _better_ compression ration and _fewer_ GB written. Why is the image larger then? Is h070 something you made with special parameters?

**Reply by probonopd** [reply]
Will be moving to an entirely different methodology for the Live system, let's look at the sizes again once it is implemented.

**Comment by probonopd** [comment]
https://github.com/helloSystem/ISO/commit/c97e25ebe1c8f02e833306c1cd86a08791c56e11

When using the same compression as with FreeBSD 12.x, then the FreeBSD 13.0 based ISO becomes even larger, exceeding 2 GB:

```
+ THRESHOLD_BYTES=2147483647
+ stat -f%z /tmp/cirrus-ci-build/artifacts/hello-0.7.0_0G39-FreeBSD-13.0-amd64.iso
+ ISO_SIZE=2148382720
+ [ 2148382720 -gt 2147483647 ]
+ echo 'Size exceeds GitHub Releases file size limit; splitting the ISO'
```

https://cirrus-ci.com/task/6308002190852096

For some reason I don't quite understand, our weird zfs+zfscompression+uzip combination apparently performs worse under FreeBSD 13 than under 12.x?

**Comment by probonopd** [comment]
The FreeBSD 13 based experimental image is unbootable after installing:
https://github.com/helloSystem/hello/issues/242

No clue why.
